Do some stuff that was inspired by linking checkboxes, but ended up not linking anything

loidas opened the pull request:
Just a random collection of tweaks.

~~I made a thing. I am not really sure myself whether it is a thing one would want to have. What I did is add some jQuery code that uses a data attribute to basically link multiple checkboxes together. If one of the linked boxes is checked, all of them will be checked; if one is unchecked, all will be unchecked. To use it, you have to set the data-link attributes on the involved entries (using the same names as for the existing data-id). [Of course this makes the HTML even more unreadable]~~

~~One usage scenario is when you have two checkboxes in different sections of the Playthrough that correspond to the same event. (Look at the Greirat and Firelink changes to see what I mean.) With this code you only have to check one box to mark the event as completed in both places. Another usage scenario is as follows: You do something in the game that qualifies for an achievement (e.g., find a covenent, defeat a boss, or reach an ending). You tick the checkbox in the Playthrough tab, while the code automatically marks the corresponding box in the Achievements tab as checked as well.~~

~~As far as backwards compatibility is concerned, the implementation should be such that if someone currently has one of the linked boxes checked and another one unchecked, nothing will change once the code is added. Only in the future when one of the linked boxes is toggled the other ones will by synced up.~~

~~But in general I wonder whether the whole thing can even provide a good user experience at all. The procedure is completely intransparent since there is no indication to the user, neither with respect to checkboxes being able to affect one another in principle nor to which affect which in practice. Not sure whether one would want that. I think my choices what to link are reasonably sane and should not be too surprising. But maybe there are some users who want to have some bosskill boxes checked but keep the corresponding achievement unchecked on purpose, or the other way round, or whatever. I have no idea. What do you think? You can try it out at https://loidas.github.io/dark-souls-3-cheat-sheet/~~


~~Other questions:~~
~~- What's the policy on adding new nodes that don't refer to any additional game content but just try to improve the presentation of the content that is already there? I'm worried that everytime I add a new box I break someone's 100% playthrough and turn it into one where not everything is marked as completed anymore since the new box starts empty.~~
~~- What's the policy on whitespace changes, especially when mixed in with content changes? I made quite a few of those where I replaced tabs with spaces in the code surrounding my changes. [Thus the code becomes a bit more readable in turn] All the tabs seem quite out of place and mess up the code alignment.~~

loidas committed:
Implement linked checkboxes

loidas committed:
Link the Endings

loidas committed:
Link the Bosses

loidas committed:
Rearrange Greirat questline

They way in which Greirat's questline is currently presented in the sheet is somewhat suboptimal:
- It tells you to free him from his cell before you can even have the key.
- It tells you that his gesture is gotten later, while it can actually be obtained immediately after presenting the bone.
- It tells you to send him away right after killing a boss, making it a long stretch before he returns with the next boss kill.
- It tells you about sending him to his death long before his ashes can be retrieved. If he is sent off at this point, his wares will be unavailable for a long time.

Some of this is improved by moving and rewording some entries. But I cannot, for example, simply remove an existing checkbox from the beginning of Road of Sacrifices and add one to the end of Undead Settlement instead, as that would mess up the counter for Road of Sacrifices. Therefore, I add some new entries at the points where I think this questline should be advanced, and link them to the existing checkboxes.

loidas committed:
Link Firelink Shrine duplicate entries

... at least the ones that are straightforward to link. The others don't really correspond one-to-one.

loidas committed:
Link some Misc Achievements

loidas committed:
Link Covenant Achievements

loidas committed:
Rearrange Anri questline

... you can get Anri's Straight Sword right after the Rite of Avowal. Similar to the situation with Greirat the checkbox can't simply be moved from Anor Londo to Irithyll of the Boreal Valley. Once again I use this to demonstrate the linking feature by adding a new linked checkbox to Irithyll of the Boreal Valley.

loidas commented:
Unrelated question:
In the function that I modify, what is the purpose of the selector `$(this).parent().parent().find('li > label > input[type="checkbox"]')`? As far as I can tell, all checkboxes are added as a `div` containing a `label` containing the `input`, Therefore, `input`'s `parent`'s `parent` never contains `li` (and even if it did, that `li` wouldn't have `label` as a direct child). So what's the deal here? Is that just some dead code or is it actually supposed to run somewhere?

ZKjellberg commented:
Hi @loidas , lot to process here.

**1. Linking checkboxes**

The Javascript code to sync items in the checklist is good. Fairly simple and it works well for bosses & endings, but I worry about the scope of introducing a feature like this as the next logical step would be to link all gear pickups with their associated page, meaning the `data-link` will quickly get a bit crazy. (Picking up an armor set associates with 4 items, and then any additional gear in that bullet)

Additionally, as you highlighted, it isn't very intuitive to the user that the other tabs are being synchronized without adding a notification would get annoying fast. This is a challenge as I think this is a good contribution, but I think it could get really messy fast. Potentially if we add, I might try to restrict its use case.

Reviewing the code, I was confused a few times. [For example, why Gundyr was linked to two different related objects.](https://github.com/ZKjellberg/dark-souls-3-cheat-sheet/pull/183/files#diff-eacf331f0ffc35d4b482f1d15a887d3bR1517) until I realized killing Gundyr would help trigger this goal. This is efficient, but again adds to the complexity which is worrying.

Additionally, looking over the Greirat link, I was tremendously confused. Even knowing the quest that related, I had to cheat at the source code and see what items linked to figure out you have a Lothric Castle bullet checking off a Grand Archives bullet. As I normally use the guide with **Hide Completed**, I would have no idea I checked off a Grand Archives bullet as I usually like to 100% a zone before moving on, even if I didn't really complete the task.

**2. New Nodes, rearranging nodes**

Regarding the addition of **new nodes**, there are no issues. The main problem is when we start moving the nodes out of order (which retains user's state) as you did with Greirat's bullet [data-18](https://github.com/ZKjellberg/dark-souls-3-cheat-sheet/pull/183/files#diff-eacf331f0ffc35d4b482f1d15a887d3bL211) to a different location. As long as all data nodes exist, the checklist will function for that zone.

The problem is maintainability of those sections as moving node numbers around can cause a headache for predicting the next node number, as we often encountered duplicate entries or skipped numbers. For example, when you added a new Greirat bullet, I was really confused trying to figure out where **playthrough_18_58** was to justify **playthrough_18_59** until I finally found it at the front of the list due to an old edit of mine. If you add a new value skipping a number, the checklist will halt, assuming the previous entry was the final.

Since the guide has mellowed in usage, I'm okay with redoing zones to a degree. We've revamped zones multiple times, so I'm open to change. Additionally, if your goal is just to rearrange, I'm okay with this. I was thinking the potentially we should add comments for the heavily re-ordered sections to note the next value, but then at that point I'd consider just re-numbering that zone.

**Whitespace**

Whitespace is something I've tried to clean up. Tabs really need to be removed and have the document properly formatted. The problem was early on the code was extremely active so I wanted to avoid merge conflicts just because I wanted to do a formatting cleanup. I'm open to finally doing a code formatting cleanup after we resolve these issues. And we implement it before you find more to fix. :P

I've used **prettify** in the past for code formatting and I remember it caused some serious problems with the code as checkboxes were somehow being cut off. The easy fix was just to revert it and revisit later. Especially with how big this document has become, it'll be a small challenge to implement the change and make sure the code hasn't been broken.

**JAVASCRIPT**

The code for this project is fairly old and has had lots of hands involved. Reviewing the original project I started this from, [it is indeed from the original source](https://github.com/smcnabb/dark-souls-2-cheat-sheet/blob/gh-pages/js/main.js#L33). That Javascript file has become out of control and I happily welcome any spring cleaning to remove dead code.

loidas committed:
Remove dead code

loidas committed:
Yoel/Yuria questline

- Added a reminder in Farron Keep to Draw Out True Strength with Yoel 5 times before Abyss Watchers and linked it to the node where Yuria is originally mentioned during Undead Settlement.
- Also fixed some small inaccuracies in events relating to that questline

ZKjellberg commented:
@loidas To follow up on this PR. I do not think I want to merge this linked functionality due to the topics we discussed. While I think the work you did is great, I think it creates some unusual behaviors and opens up challenges.

I'd be happy to merge the rest of your tweaks. Additionally, we can explore remove tabs and reformatting the document, but I'd like to complete this task at a time that will not cause merge conflict challenges.

loidas committed:
Orbeck questline

- Moved the reminder about giving him scrolls in Farron Keep to a point where you might be in Firelink Shrine anyway ( or at least close to a bonfire)
- Added a new checkbox right afterwards about getting rewards from him for turning in scrolls and buying spells, now that you can actually get them after giving a scroll. Linked it to the node where rewards are first mentioned.
- Clarified the conditions for his rewards in the checklists
- Renamed Pestilent Mercury to Pestilent Mist

loidas committed:
Remove linked checkboxes

ZKjellberg commented:
@loidas Is ready for merging?

loidas commented:
If you are okay with these changes, then yes.

Minor sidenote:
If you know a good way to hyperlink from one tab to another (or even better, if you know a way to link to a specific section on the other tab), feel free to improve my solution. Currently I use `<a href="#" onclick="$('.nav.navbar-nav li a[href=\x22#tabMisc\x22]').click();">Misc</a>` to go to the top of the Misc tab. I have no idea how to point directly to the pumparum section, which is where these cross references should ideally be going.

ZKjellberg commented:
Merged. Not sure on how to improve the linking to content within another tab, but functionality you implemented works well.

ZKjellberg commented:
I've updated the code, removing all tabs. Thanks for reporting this. I'm looking into running [atom-beautify](https://atom.io/packages/atom-beautify) or **Reformat Code** type tool to improve readability when reviewing source.

IntelliJ's **Reformat Code** has an issue where format code forces entire document to have a page width, which causes lots of problems with <span> elements. I'll keep you updated.

ZKjellberg commented:
Just a heads up. I'm toying with [atom-beautify](https://atom.io/packages/atom-beautify) but it is causing some really crazy formatting due to the long length of some bullets.

Additionally, this guide is missing many </li> tags which confuse the formatting even more. I'm somewhat shocked the W3C Validator never caught these.

Expect a large commit soon which will effect most of the content, so if you are looking into any further updates, expect conflicts until I push the changes.

ZKjellberg commented:
Okay. All commits have been pushed. This was actually much more minor than I expected.

I originally tried to do a massive revamp to reduce the document width, but if you create a newline in source, it causes the <li> tag to end prematurely, ignoring all content after the line break. Good news is that going through this process allowed me to find lots of missing </li> tags.

Code is now open for improvement if you have further ideas.

Thanks again for the contributions. 👍

loidas commented:
You didn't have to worry about the missing `</li>` tags, omitting them is perfectly valid in most cases: [An li element's end tag may be omitted if the li element is immediately followed by another li element or if there is no more content in the parent element.](https://www.w3.org/TR/html5/syntax.html#syntax-tag-omission)
